### PR TITLE
Ensure that _GNU_SOURCE is defined for NI_MAXHOST and NI_MAXSERV

### LINKED ISSUE #13049: Building OpenSSL 1.1.1h with gcc9 on linux creates build error in b_addr.c
We recently performed our own OpenSSL build based on the 1.1.1h sources from this repo as static library and ensured that we have the same compiler settings used for the build as we use elsewhere for our C++ based code. This means that we also activated ``_POSIX_C_SOURCE=200809L`` while doing so. With these settings everything builds fine except the file crypto/bio/b_addr.c, which gives us the following diagnostics for gcc9:

 ```
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c: In function ‘addr_strings’:
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:198:19: error: ‘NI_MAXHOST’ undeclared (first use in this function)
[INFO]   198 |         char host[NI_MAXHOST] = "", serv[NI_MAXSERV] = "";
[INFO]       |                   ^~~~~~~~~~
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:198:19: note: each undeclared identifier is reported only once for each function it appears in
[INFO] [ 15%] Building C object crypto/CMakeFiles/crypto.dir/bio/bf_null.c.o
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:198:42: error: ‘NI_MAXSERV’ undeclared (first use in this function)
[INFO]   198 |         char host[NI_MAXHOST] = "", serv[NI_MAXSERV] = "";
[INFO]       |                                          ^~~~~~~~~~
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:198:37: warning: unused variable ‘serv’ [-Wunused-variable]
[INFO]   198 |         char host[NI_MAXHOST] = "", serv[NI_MAXSERV] = "";
[INFO]       |                                     ^~~~
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:198:14: warning: unused variable ‘host’ [-Wunused-variable]
[INFO]   198 |         char host[NI_MAXHOST] = "", serv[NI_MAXSERV] = "";
[INFO]       |              ^~~~
[INFO] [ 15%] Building C object crypto/CMakeFiles/crypto.dir/bio/bio_cb.c.o
[INFO] In file included from /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/include/internal/cryptlib.h:25,
[INFO]                  from /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/bio_local.h:89,
[INFO]                  from /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:13:
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c: In function ‘BIO_lookup_ex’:
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:811:52: error: ‘h_errno’ undeclared (first use in this function); did you mean ‘errno’?
[INFO]   811 |                 SYSerr(SYS_F_GETHOSTBYNAME, 1000 + h_errno);
[INFO]       |                                                    ^~~~~~~
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/include/openssl/err.h:29:61: note: in definition of macro ‘ERR_PUT_error’
[INFO]    29 | #  define ERR_PUT_error(a,b,c,d,e)        ERR_put_error(a,b,c,d,e)
[INFO]       |                                                             ^
[INFO] /home/jenkins/jenkins/workspace/dmb1/com.github.openssl-linux-x86_64-gcc9/src/crypto/bio/b_addr.c:811:17: note: in expansion of macro ‘SYSerr’
[INFO]   811 |                 SYSerr(SYS_F_GETHOSTBYNAME, 1000 + h_errno);
[INFO]       |                 ^~~~~~
[INFO] make[2]: *** [crypto/CMakeFiles/crypto.dir/build.make:1419: crypto/CMakeFiles/crypto.dir/bio/b_addr.c.o] Error 1
```
The errors all can be attributed to the same cause:

* Regarding ``NI_MAXHOST``: According to [the man pages](https://man7.org/linux/man-pages/man3/getnameinfo.3.html):
> Since glibc 2.8, these definitions are exposed only if suitable feature test macros are defined, namely: _GNU_SOURCE, 
> _DEFAULT_SOURCE (since glibc 2.19), or (in glibc versions up to and including 2.19) _BSD_SOURCE or _SVID_SOURCE.
* Regarding `` h_errno``: There seem to exist similar need to add ``_GNU_SOURCE``

We could resolve all compile problems by applying the following changes:
```
--- src/crypto/bio/b_addr.c	
+++ src/crypto/bio/b_addr.c	
@@ -7,6 +7,10 @@
  * https://www.openssl.org/source/license.html
  */
 
+#ifndef _GNU_SOURCE
+# define _GNU_SOURCE
+#endif
+
 #include <assert.h>
 #include <string.h>
 ```
This fix seems to be in sync with some other translation units of OpenSSL which also conditionally define ``_GNU_SOURCE`` at the very beginning.

### PR DESCRIPTION
**Pull request by Dani-Hub**
This fixes (among others) the following error with gcc9:

.../crypto/bio/b_addr.c:198:19: error: ‘NI_MAXHOST’ undeclared

Since glibc 2.8, these defines like `NI_MAXHOST` are exposed only
if suitable feature test macros are defined, namely: _GNU_SOURCE,
_DEFAULT_SOURCE (since glibc 2.19), or _BSD_SOURCE or _SVID_SOURCE
(before glibc 2.19)

CLA: trivial
Fixes #13049

### CODE REVIEW

**Comment by kroeckx** [comment]
_GNU_SOURCE is probably the best to define, even though we don't
need any gnu feature. _BSD_SOURCE and _SVID_SOURCE have been
deprecated in favour of _DEFAULT_SOURCE and generate a warning,
and _DEFAULT_SOURCE doesn't work with older releases.

**Comment by mattcaswell** [comment]
Is this only relevant for master? Or for 1.1.1 as well?

**Comment by Dani-Hub** [comment]
> 
> 
> Is this only relevant for master? Or for 1.1.1 as well?

I had been advised to make a pull request to master. I guess that limits the fix for OpenSSL 3.0, right? Personally I would like to see it in 1.1.1, which is relevant for us. Would I be supposed to make a second pull request then and if so, to which exact branch?

**Comment by mattcaswell** [comment]
If it can be cherry-picked cleanly from one branch to the other then we can just use this PR for both.

**Review by mattcaswell — APPROVED** [review]
Looks ok to me.

**Comment by Dani-Hub** [comment]
> 
> 
> If it can be cherry-picked cleanly from one branch to the other then we can just use this PR for both.

I'm not so experienced with git to ensure that I could perform a cherry-pick, so if it is also OK, that I provide a second pull request, I would prefer to follow that route. Regardless of the mechanism: What would be the correct branch that I should address? Would "OpenSSL_1_1_1-stable" be the right one?

**Comment by mattcaswell** [comment]
> I'm not so experienced with git to ensure that I could perform a cherry-pick, so if it is also OK, that I provide a second pull request, I would prefer to follow that route. Regardless of the mechanism: What would be the correct branch that I should address? Would "OpenSSL_1_1_1-stable" be the right one?

I just checked and it does cherry-pick cleanly so there is no need for a second PR. It is the responsibility of the committer (not you) to do the actual cherry-picking when committing anyway.

**Comment by Dani-Hub** [comment]
> 
> 
> > I'm not so experienced with git to ensure that I could perform a cherry-pick, so if it is also OK, that I provide a second pull request, I would prefer to follow that route. Regardless of the mechanism: What would be the correct branch that I should address? Would "OpenSSL_1_1_1-stable" be the right one?
> 
> I just checked and it does cherry-pick cleanly so there is no need for a second PR. It is the responsibility of the committer (not you) to do the actual cherry-picking when committing anyway.

Thanks for your help!

**Comment by mattcaswell** [comment]
Needs a second reviewer.

**Review by levitte — APPROVED** [review]

**Comment by mspncp** [comment]
Editorial note: I think that the reference to `_POSIX_C_SOURCE=200809L` in the commit message

> ```
> Ensure that _GNU_SOURCE is defined if _POSIX_C_SOURCE=200809L is defined
> ```

is a little bit misleading, because the patch does not introduce any version check. (It should actually be a `>=`, see [getnameinfo(3)](https://man7.org/linux/man-pages/man3/getnameinfo.3.html)). After a little investigation, I would suggest a more general formulation like the following:

```
Ensure that _GNU_SOURCE is defined for NI_MAXHOST and NI_MAXSERV

Since glibc 2.8, these defines like `NI_MAXHOST` are exposed only
if suitable feature test macros are defined, namely: _GNU_SOURCE,
_DEFAULT_SOURCE (since glibc 2.19), or _BSD_SOURCE or _SVID_SOURCE
(before glibc 2.19), see GETNAMEINFO(3).

CLA: trivial
Fixes #13049
```

(@Dani-Hub if you are not familiar with amending the commit message without opening a new pull request you don't need to do it yourself. Just drop us a note that you agree with the change, then one of us can do it when we merge the pr.)

**Comment by Dani-Hub** [comment]
> Editorial note: I think that the reference to `_POSIX_C_SOURCE=200809L` in the commit message
> 
> > ```
> > Ensure that _GNU_SOURCE is defined if _POSIX_C_SOURCE=200809L is defined
> > ```
> 
> is a little bit misleading, because the patch does not introduce any version check. (It should actually be a `>=`, see [getnameinfo(3)](https://man7.org/linux/man-pages/man3/getnameinfo.3.html)). After a little investigation, I would suggest a more general formulation like the following:
> 
> ```
> Ensure that _GNU_SOURCE is defined for NI_MAXHOST and NI_MAXSERV
> 
> Since glibc 2.8, these defines like `NI_MAXHOST` are exposed only
> if suitable feature test macros are defined, namely: _GNU_SOURCE,
> _DEFAULT_SOURCE (since glibc 2.19), or _BSD_SOURCE or _SVID_SOURCE
> (before glibc 2.19), see GETNAMEINFO(3).
> 
> CLA: trivial
> Fixes #13049
> ```

Yes, I agree with that change - Thanks for taking care of the update!

**Comment by richsalz** [comment]
Isn't this adding a new platform?  I mean yes, it's trivial, but doesn't it go against the OS390 decision, and the pending Apple hardware support decision?

**Comment by mspncp** [comment]
> Isn't this adding a new platform? I mean yes, it's trivial, but doesn't it go against the OS390 decision, and the pending Apple hardware support decision?

No, I don't think so. It's glibc specific, not platform specific.

**Review by mspncp — APPROVED** [review]
Approved with the new commit message.

(@mattcaswell or @levitte please confirm the change of the commit message. Therefore, I'll take care of the merge.)

**Comment by mspncp** [comment]
Oh, and for the records: I agree that it's a (legally) trivial change.

**Comment by mattcaswell** [comment]
Reconfirmed. I agree this is a trivial change.

**Comment by mspncp** [comment]
Just changed the commit message to the one suggested, [without tree changes](https://github.com/openssl/openssl/compare/e7372810d5bc9feff822f72b40df55d133c0b0f8..aee2f50a3e49d53597755828bd23a94e795e8bc7). No reapproval required.

**Comment by mspncp** [comment]
Merged to master and 1.1.1. Thank you @Dani-Hub and congratulations to your first contribution to OpenSSL! :-)

* 99501e817c Ensure that _GNU_SOURCE is defined for NI_MAXHOST and NI_MAXSERV  (master)
* db78f8b385 Ensure that _GNU_SOURCE is defined for NI_MAXHOST and NI_MAXSERV  (1.1.1)

**Pull request closed without merging**
